R/kmeans-clustering-method.R

```diff
@@ -0,0 +1,528 @@
+#' ## or represented by each principle component.
+#' cl$getScreePlot()
+#' 
+#' # According to the scree plot, we decide to use the first 3 PCs to do clustering.
+#' p$usePrinComp <- TRUE
+#' p$numOfPrinComp <- 3
+#' 
+#' # Run k means clustering
+#' cl <- KmeansClustering$new(p)
+#' cl$run()
+#' 
+#' # Get the fit result
+#' cl$getKmeansfit()
+#'
+#' @export
+
+KmeansClustering <- R6Class("KmeansClustering",
+                            
+  # Inheritance
+  inherit = UnsupervisedModel,
+  
+  # Private members
+  private = list(
+    # Get kmeans model
+    kmeans.fit = NA,
+    confusionMatrix = NA,
+    vnum = NA,
+    prin.comp = NA,
+    df.prin = NA,
+    prop_varex = NA,
+    centers = NA,
+    cluster = NA,
+    outDf = NA,
+    cluster.labels = NA,
+    mean.vec = NA,
+    sd.vec = NA,
+    scaledf = NA,
+
+    
+    #check if the data type is numeric
+    checkDataType = function() {
+      #print(self$params$dataType)
+      if (self$params$dataType != 'numeric' || isNumeric(self$params$df) == FALSE) {
+        stop("Your data type must be numeric in order to use kmeans.")
+      }
+    },
+    
+    # Scale the data
+    dataScale = function() {
+      if (isTRUE(self$params$debug)) {
+        print('scale the data set for k means clustering...')
+      }
+      private$mean.vec <- apply(self$params$df, 2, mean, na.rm = TRUE)
+      private$sd.vec <- apply(self$params$df, 2, sd, na.rm = TRUE)
+      r <- nrow(self$params$df)
+      c <- ncol(self$params$df)
+      private$scaledf <- matrix(data = NA,nrow = r, ncol = c)
+      private$scaledf <- data.frame(private$scaledf)
+      names(private$scaledf) <- names(self$params$df)
+      for (i in 1:r) {
+        private$scaledf[i,] <- (self$params$df[i,] - private$mean.vec)/private$sd.vec
+      }
+    },
+    
+    # PCA analysis
+    pcaAnalysis = function() {
+      if (isTRUE(self$params$debug)) {
+        print('performing PCA analysis...')
+      }
+      private$prin.comp <- prcomp(self$params$df, scale. = T)
+      #compute standard deviation of each principal component
+      std_dev <- private$prin.comp$sdev
+      #compute variance
+      pr_var <- std_dev^2
+      #proportion of variance explained
+      private$prop_varex <- pr_var/sum(pr_var)
+    },
+    
+    
+    # Extract the principle components 
+    prinCompDf = function() {
+      if (isTRUE(self$params$debug)) {
+        print('extracting principle components...')
+      }
+      private$df.prin <- data.frame(private$prin.comp$x)
+      private$df.prin <- private$df.prin[,1:self$params$numOfPrinComp]
+      
+      if (isTRUE(self$params$debug)) {
+        print('The principle components generated by PCA...')
+        print(str(private$df.prin))
+      }
+    },
+    
+    
+    # The following helper functions were found at
+    # paulbourke.net/geometry/pointlineplane/pointline.r
+    # via the SO reference below.  The short segment check
+    # was modified to permit vectorization.
```

Comment:
As was mentioned, can we move this link to a common function (and also clean up extra blank lines here)?